File: 0/0/2/002.py
# Copyright (C) 2013 Jorge Aparicio


def fibonacci(n):
    if n not in fib:
        fib[n] = fib[n - 1] + fib[n - 2]
    return fib[n]


def even_fib_up_to(n):
    i = 1
    while True:
        f = fibonacci(i)
        if f > n:
            break
        if f % 2 == 0:
            yield f
        i += 1

fib = {1: 1, 2: 2}
print(sum(i for i in even_fib_up_to(4e6)))

File: 0/0/2/002.rs
// Copyright (C) 2013 Jorge Aparicio

use std::iter::AdditiveIterator;
use std::util::replace;

fn main() {
    println!("{}",
        fibonacci().take_while(|&x| x < 4000000).filter(|x| x % 2 == 0).sum());
}

struct Fibonacci { curr: uint, next: uint }

fn fibonacci() -> Fibonacci {
    Fibonacci { curr: 1, next: 2 }
}

impl Iterator<uint> for Fibonacci {
    fn next(&mut self) -> Option<uint> {
        let new_next = self.curr + self.next;
        let new_curr = replace(&mut self.next, new_next);
        Some(replace(&mut self.curr, new_curr))
    }
}

